src/convert_image/x86.rs

```diff
@@ -428,6 +428,127 @@ fn lrgb_to_i420(
+    channels: PixelFormatChannels,
+    colorimetry: Colorimetry,
+    sampler: Sampler,
+) -> bool {
+    if last_dst_plane != 2
+        || (last_dst_plane >= dst_strides.len())
+        || (last_dst_plane >= dst_buffers.len())
+        || src_strides.is_empty()
+        || src_buffers.is_empty()
+    {
+        return false;
+    }
+
+    let depth = channels as usize;
+    let col_count = width as usize;
+    let line_count = height as usize;
+    let packed_rgb_stride = depth * col_count;
+
+    let rgb_stride = if src_strides[0] == 0 {
+        packed_rgb_stride
+    } else {
+        src_strides[0]
+    };
+
+    let y_stride = if dst_strides[0] == 0 {
+        col_count
+    } else {
+        dst_strides[0]
+    };
+
+    let u_stride = if dst_strides[1] == 0 {
+        col_count
+    } else {
+        dst_strides[1]
+    };
+
+    let v_stride = if dst_strides[2] == 0 {
+        col_count
+    } else {
+        dst_strides[2]
+    };
+
+    let rgb_plane = &src_buffers[0];
+    let (y_plane, uv_plane) = dst_buffers.split_at_mut(1);
+    let (u_plane, v_plane) = uv_plane.split_at_mut(1);
+
+    let y_plane = &mut y_plane[0][..];
+    let u_plane = &mut u_plane[0][..];
+    let v_plane = &mut v_plane[0][..];
+
+    if line_count == 0 {
+        return true;
+    }
+
+    let max_stride = usize::max_value() / line_count;
+    if (y_stride > max_stride)
+        || (u_stride > max_stride)
+        || (v_stride > max_stride)
+        || (rgb_stride > max_stride)
+    {
+        return false;
+    }
+
+    if y_stride * line_count > y_plane.len()
+        || u_stride * line_count > u_plane.len()
+        || v_stride * line_count > v_plane.len()
+        || rgb_stride * line_count > rgb_plane.len()
+    {
+        return false;
+    }
+
+    // The following constants will be automatically propagated as immediate values
+    // inside the operations by optimizing compilers.
+    let col = colorimetry as usize;
+    let xr = FORWARD_WEIGHTS[col][0];
+    let xg = FORWARD_WEIGHTS[col][1];
+    let xb = FORWARD_WEIGHTS[col][2];
+    let yr = FORWARD_WEIGHTS[col][3];
+    let yg = FORWARD_WEIGHTS[col][4];
+    let yb = FORWARD_WEIGHTS[col][5];
+    let zr = FORWARD_WEIGHTS[col][6];
+    let zg = FORWARD_WEIGHTS[col][7];
+    let zb = FORWARD_WEIGHTS[col][8];
+    unsafe {
+        let rgb_group = rgb_plane.as_ptr();
+        let y_group = y_plane.as_mut_ptr();
+        let u_group = u_plane.as_mut_ptr();
+        let v_group = v_plane.as_mut_ptr();
+
+        for y in 0..line_count {
+            for x in 0..col_count {
+                let (r, g, b) = unpack_ui8x3_i32(
+                    rgb_group.add(wg_index(x, y, depth, rgb_stride)),
+                    sampler,
+                );
+
+                let y_data = y_group.add(wg_index(x, y, 1, y_stride));
+                *y_data = fix_to_i32(affine_transform(r, g, b, xr, xg, xb, Y_OFFSET), FIX16) as u8;
+
+                let u_data = u_group.add(wg_index(x, y, 1, u_stride));
```

Comment:
Here, chroma calculation uses `FIX18`, but since we are not averaging 2x2 neighbourhood, we should use `FIX16`. This affects also the computation of `C_OFFSET`, which has to be computed in 16.16 fixed point

Reply:
fixed